pkg/imgpkg/v1/pull.go

```diff
@@ -0,0 +1,207 @@
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/vmware-tanzu/carvel-imgpkg/pkg/imgpkg/bundle"
+	"github.com/vmware-tanzu/carvel-imgpkg/pkg/imgpkg/plainimage"
+	"github.com/vmware-tanzu/carvel-imgpkg/pkg/imgpkg/registry"
+)
+
+// Logger Interface used for logging
+type Logger interface {
+	Errorf(msg string, args ...interface{})
+	Warnf(msg string, args ...interface{})
+	Debugf(msg string, args ...interface{})
+	Tracef(msg string, args ...interface{})
+	UI() goui.UI
+}
+
+// PullOpts Option that can be provided to the pull request
+type PullOpts struct {
+	Logger Logger
+	// AsImage Pull the contents of the OCI Image
+	AsImage bool
+	// IsBundle the image being pulled is a Bundle
+	IsBundle bool
+}
+
+// ImagesLockInfo Information about the ImagesLock file
+type ImagesLockInfo struct {
+	Path    string `json:"path"`
+	Updated bool   `json:"updated"`
+}
+
+// BundleInfo Information related to the specific bundle
+type BundleInfo struct {
+	ImageRef      string          `json:"image"`
+	ImagesLock    *ImagesLockInfo `json:"imagesLock,omitempty"`
+	NestedBundles []BundleInfo    `json:"inner-bundles,omitempty"`
+}
+
+// InfoFromNestedBundle Given a Bundle struct it can generate the BundleInfo associated to the nested bundles
+func InfoFromNestedBundle(b *bundle.Bundle, updated bool) []BundleInfo {
+	var nestedBundles []BundleInfo
+
+	for _, nBundle := range b.NestedBundles() {
+		nestedBundles = append(nestedBundles, NewBundleInfo(nBundle, updated))
+	}
+	return nestedBundles
+}
+
+// NewBundleInfo create a BundleInfo struct from bundle information
+func NewBundleInfo(node bundle.GraphNode, updated bool) BundleInfo {
+	bInfo := BundleInfo{
+		ImageRef: node.ImageRef,
+		ImagesLock: &ImagesLockInfo{
+			Path:    filepath.Join(node.Path, bundle.ImgpkgDir, bundle.ImagesLockFile),
+			Updated: updated,
+		},
+		NestedBundles: nil,
+	}
+	for _, nestedBundle := range node.NestedBundles {
+		bInfo.NestedBundles = append(bInfo.NestedBundles, NewBundleInfo(nestedBundle, updated))
+	}
+	return bInfo
+}
+
+// Status Report from the Pull command
+type Status struct {
+	BundleInfo
+	IsBundle  bool `json:"-"`
+	Cacheable bool `json:"cacheable"`
+}
+
+// Pull Download the contents of the image referenced by imageRef to the folder outputPath
+func Pull(imageRef string, outputPath string, pullOptions PullOpts, registryOpts registry.Opts) (Status, error) {
+	reg, err := registry.NewSimpleRegistry(registryOpts)
+	if err != nil {
+		return Status{}, err
+	}
+
+	isBundle, err := bundle.NewBundle(imageRef, reg).IsBundle()
+	if err != nil {
+		return Status{}, err
+	}
+
+	switch {
+	case isBundle && pullOptions.AsImage: // Trying to pull the OCI Image of a Bundle
+		st, err := pullImage(imageRef, outputPath, pullOptions, reg)
+		if err != nil {
+			return Status{}, err
+		}
+		st.IsBundle = true
+		return st, nil
+
+	case isBundle && pullOptions.IsBundle: // Trying to pull a Bundle
+		return pullBundle(imageRef, outputPath, pullOptions, reg, false)
```

Comment:
it seems like we are now making an extra registry call to check if something is a bundle... bundle.NewBundle().IsBundle and we build another copy of the bundle inside pullBundle.